lib/web/apiserver.go

```diff
@@ -1751,10 +1751,91 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
 			log.Warn(err)
 			return nil, trace.Wrap(err)
 		}
+
 		return fn(w, r, p, ctx, site)
 	})
 }
 
+// clientWithUserRole will return an auth.ClientI with the role of the user at
+// the requested site. If the site is local a client with the users local role
+// is returned. If the site is remote a client with the users remote role is
+// returned.
+func clientWithUserRole(ctx *SessionContext, site reversetunnel.RemoteSite) (auth.ClientI, error) {
+	// get the name of the current cluster
+	lclt, err := ctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cn, err := lclt.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// if we're trying to access the local cluster, pass back the local client.
+	if cn.GetClusterName() == site.GetName() {
+		return lclt, nil
+	}
+
+	// we're trying to access a remote cluster, get a client with the role this
+	// user will assume on the remote cluster.
+	clt, err := remoteClient(ctx, site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return clt, nil
+}
+
+// remoteClient returns a client to a remote site with the role of logged in user.
+func remoteClient(ctx *SessionContext, site reversetunnel.RemoteSite) (auth.ClientI, error) {
+	sshDialer := func(network, addr string) (net.Conn, error) {
+		// we tell the remote site we want a connection to the auth server by using
+		// a non-resolvable string @remote-auth-server as the destination.
+		srcAddr := utils.NetAddr{Addr: "web.get-remote-client-with-user-role", AddrNetwork: "tcp"}
+		dstAddr := utils.NetAddr{Addr: reversetunnel.RemoteAuthServer, AddrNetwork: "tcp"}
+
+		// first get a net.Conn (tcp connection) to the remote auth server. no
+		// authentication has occured.
+		netConn, err := site.Dial(&srcAddr, &dstAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// get the principal and authMethods we will use to authenticate on the
+		// remote cluster from our local cluster
+		agent, err := ctx.GetAgent()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		principal, authMethods, err := getUserCredentials(agent)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// build a ssh connection to the remote auth server
+		config := &ssh.ClientConfig{
+			User:    principal,
+			Auth:    []ssh.AuthMethod{authMethods},
+			Timeout: defaults.DefaultDialTimeout,
+		}
+		sshConn, sshChans, sshReqs, err := ssh.NewClientConn(netConn, reversetunnel.RemoteAuthServer, config)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		client := ssh.NewClient(sshConn, sshChans, sshReqs)
+
+		// dial again to the HTTP server
+		return client.Dial(network, addr)
+	}
+
+	clt, err := auth.NewClient("http://stub:0", sshDialer)
```

Comment:
you will be creating clients on every request. This client should be associated with active user session and should be deleted when the session closes. Extend the logic of the `SessionContext` to support additional use case.